### DISCUSSION TITLE
AzureRM v4.0 thread

### DISCUSSION THREAD

**Discussion by matt-FFFFFF** [Azure/Azure-Verified-Modules#1338] [General]
## AzureRM v4 guidance for Azure Verified Modules

- All modules must support v4.x by end of January 2025
- Before January 2025, module authors should update their modules based on demand for new features and release a new version
- Module authors should try to maintain backwards compatibility with v3 until new features or quality improvements in v4 mean this no longer possible
- Until January 2025, new module should support both v3 and v4 unless features or resource usage demand only v4
- Referenced module versions must be pinned as per the spec.

**updated** with guidance summary 2024-09-11

**Comment by matt-FFFFFF** [comment]
I should mention that the initial AVM guidance will be released 1st week Sept

**Comment by ryan-royals** [comment]
I have only done a little tinkering with 4.0 in our environments, but I don't see any reason to not move forward with it.
The changes to the provider config are pretty minimal (Being needing to supply the Sub ID), and this has been our standard practise forever, and I assume its the same for lots of people.

Beyond that, a lot of the breaking changes for the resources don't appear to be that day to day frustrating beyond the `log` >> `enabled_log` for `azurerm_monitor_diagnostic_setting`, but this one has been warned about for so long that I think its reasonable to assume that most users can figure out what has happened.

Perhaps if for AVM there was a standard approach to this?
Patch releases for provider are with next update (Unless directly effects module).
Minor releases are release + 7 days.
Major releases are release + 14 days.

**Reply by matt-FFFFFF** [reply]
We've updated our guidance above - would love your feedback

**Reply by ryan-royals** [reply]
(Sorry was on holidays 🍸)

Having some dates is perfect in my mind, seems simple enough to follow and enforce.

**Comment by alefteris** [comment]
There should also be a check that AVM modules don't have an upper bound restriction on the azurerm provider version, unless there is a specific reason for it. See for example https://github.com/Azure/terraform-azurerm-avm-res-containerregistry-registry/issues/79 that restricts to azurerm < 4.0.0.

**Reply by matt-FFFFFF** [reply]
This isn't going to happen. The major version constraint is there to stop breaking changes where we have been unable to test future versions.

Once a test for v4 has been completed, the version constraint can be moved from < 4.0.0 to < 5.0.0. This is what's happened on the keyvault module.

**Reply by alefteris** [reply]
@matt-FFFFFF Hi. On a second thought, it makes sense to pin it to below v4 for now, you are right. I saw that for ex. the container registry needs quite a few changes to work on v4.

Question: in the key vault module I don't see any restriction to < 5. On purpose or it has been missed?

**Reply by matt-FFFFFF** [reply]
Been missed - will add 😊
